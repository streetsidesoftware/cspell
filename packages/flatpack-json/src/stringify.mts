import type { Flatpacked } from './types.mjs';

export function stringifyFlatpacked(input: Flatpacked): string {
    let result = '[\n' + JSON.stringify(input[0]);

    let prev = '';

    for (let i = 1; i < input.length; i++) {
        const next = JSON.stringify(input[i]);
        result += prev == next ? ',' : ',\n';
        result += next;
        prev = next;
    }

    result += '\n]\n';

    return result;
}
