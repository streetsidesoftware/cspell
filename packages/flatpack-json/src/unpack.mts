import assert from 'node:assert';

import {
    ArrayBasedElements,
    ArrayElement,
    BigIntElement,
    dataHeader,
    DateElement,
    Element,
    ElementType,
    Flatpacked,
    MapElement,
    ObjectElement,
    Primitive,
    PrimitiveArray,
    PrimitiveMap,
    PrimitiveObject,
    PrimitiveSet,
    RegExpElement,
    Serializable,
    SetElement,
    StringElement,
    SubStringElement,
    Unpacked,
} from './types.mjs';

export function fromJSON(data: Flatpacked): Unpacked {
    const [header] = data;

    if (header !== dataHeader) {
        throw new Error('Invalid header');
    }

    const cache = new Map<number | number[], Unpacked>([[0, undefined]]);
    /**
     * indexes that have been referenced by other objects.
     */
    const referenced = new Set<number>();

    function mergeKeysValues<K>(keys: readonly K[], values: PrimitiveArray): [K, Serializable][] {
        return keys.map((key, i) => [key, values[i]]);
    }

    function toSet(idx: number, elem: SetElement): PrimitiveSet {
        const [_, k] = elem;
        const s: PrimitiveSet = k ? (new Set(idxToArr(k)) as PrimitiveSet) : new Set();
        cache.set(idx, s);
        return s;
    }

    function toMap(idx: number, elem: MapElement): PrimitiveMap {
        const [_, k, v] = elem;
        const m: PrimitiveMap =
            !k || !v ? new Map() : (new Map(mergeKeysValues(idxToArr(k), idxToArr(v))) as PrimitiveMap);
        cache.set(idx, m);
        return m;
    }

    function toRegExp(idx: number, elem: RegExpElement): RegExp {
        const [_, pattern, flags] = elem;
        const p = idxToValue(pattern) as string;
        const f = idxToValue(flags) as string;
        const r = new RegExp(p, f);
        cache.set(idx, r);
        return r;
    }

    function toBigInt(idx: number, elem: BigIntElement): bigint {
        const [_, vIdx] = elem;
        const r = BigInt(idxToValue(vIdx) as string | number);
        cache.set(idx, r);
        return r;
    }

    function toDate(idx: number, elem: DateElement): Date {
        const [_, value] = elem;
        const r = new Date(value);
        cache.set(idx, r);
        return r;
    }

    function toString(idx: number, elem: StringElement | string): string {
        const s = typeof elem === 'string' ? elem : idxToValue(elem.slice(1) as number[]);
        cache.set(idx, s);
        return s as string;
    }

    function toObj(idx: number, elem: ObjectElement): PrimitiveObject {
        const [_, k, v] = elem;

        // Object Wrapper
        if (!k && v) {
            const obj = Object(idxToValue(v));
            cache.set(idx, obj);
            return obj as PrimitiveObject;
        }

        const obj = {};
        cache.set(idx, obj);

        if (!k || !v) return obj;
        const keys = idxToArr(k) as string[];
        const values = idxToArr(v);
        Object.assign(obj, Object.fromEntries(mergeKeysValues(keys, values)));
        return obj;
    }

    function idxToArr(idx: number): PrimitiveArray {
        const element = data[idx];
        assert(isArrayElement(element));
        return toArr(idx, element);
    }

    function toArr(idx: number, element: ArrayElement): PrimitiveArray {
        const placeHolder: Serializable[] = [];
        const refs = element.slice(1);
        cache.set(idx, placeHolder);
        const arr = refs.map(idxToValue);
        // check if the array has been referenced by another object.
        if (!referenced.has(idx)) {
            // It has not, just replace the placeholder with the array.
            cache.set(idx, arr);
            return arr;
        }
        placeHolder.push(...arr);
        return placeHolder;
    }

    function handleSubStringElement(idx: number, refs: SubStringElement): string {
        const [_t, sIdx, len, offset = 0] = refs;
        const s = `${idxToValue(sIdx)}`.slice(offset, offset + len);
        cache.set(idx, s);
        return s;
    }

    function handleArrayElement(
        idx: number,
        element: ArrayBasedElements,
    ): PrimitiveArray | Primitive | PrimitiveObject | PrimitiveSet | PrimitiveMap {
        switch (element[0]) {
            case ElementType.Array: {
                break;
            }
            case ElementType.Object: {
                return toObj(idx, element as ObjectElement);
            }
            case ElementType.String: {
                return toString(idx, element as StringElement);
            }
            case ElementType.SubString: {
                return handleSubStringElement(idx, element as SubStringElement);
            }
            case ElementType.Set: {
                return toSet(idx, element as SetElement);
            }
            case ElementType.Map: {
                return toMap(idx, element as MapElement);
            }
            case ElementType.RegExp: {
                return toRegExp(idx, element as RegExpElement);
            }
            case ElementType.Date: {
                return toDate(idx, element as DateElement);
            }
            case ElementType.BigInt: {
                return toBigInt(idx, element as BigIntElement);
            }
        }
        return toArr(idx, element as ArrayElement);
    }

    function idxToValue(idx: number | number[]): Serializable {
        if (!idx) return undefined;
        const found = cache.get(idx);
        if (found !== undefined) {
            if (typeof idx === 'number') referenced.add(idx);
            return found as Serializable;
        }

        if (Array.isArray(idx)) {
            // it is a nested string;
            const parts = idx.map((i) => idxToValue(i));
            return joinToString(parts);
        }

        const element = data[idx];

        if (typeof element === 'object') {
            // eslint-disable-next-line unicorn/no-null
            if (element === null) return null;
            if (Array.isArray(element)) return handleArrayElement(idx, element as ArrayBasedElements);
            return {};
        }
        return element;
    }

    return idxToValue(1);
}

function joinToString(parts: PrimitiveArray): string {
    return parts.map((a) => (Array.isArray(a) ? joinToString(a) : a)).join('');
}

function isArrayElement(value: Element): value is ArrayElement {
    return Array.isArray(value) && value[0] === ElementType.Array;
}

export function parse(data: string): Unpacked {
    return fromJSON(JSON.parse(data));
}
