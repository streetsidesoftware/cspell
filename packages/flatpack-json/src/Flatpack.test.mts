import { readFile } from 'node:fs/promises';

import { findMatchingFileTypes } from '@cspell/filetypes';
import { createPatch } from 'diff';
import { describe, expect, test } from 'vitest';

import { FlatpackStore, stringify, toJSON } from './Flatpack.mjs';
import { stringifyFlatpacked } from './stringify.mjs';
import { fromJSON } from './unpack.mjs';

const urlFileList = new URL('../fixtures/fileList.txt', import.meta.url);
const baseFilename = new URL(import.meta.url).pathname.split('/').slice(-1).join('').split('.').slice(0, -2).join('.');

describe('Flatpack', async () => {
    const sampleFiles = await sampleFileList();

    test.each`
        data
        ${undefined}
        ${'string'}
        ${1}
        ${1.1}
        ${null}
        ${true}
        ${false}
        ${[]}
        ${[1, 2]}
        ${['a', 'b', 'a', 'b']}
        ${{}}
        ${{ a: 1 }}
        ${{ a: { b: 1 } }}
        ${{ a: { a: 'a', b: 42 } }}
        ${{ a: [1] }}
    `('dehydrate/hydrate $data', ({ data }) => {
        const v = toJSON(data);
        expect(fromJSON(v)).toEqual(data);
    });

    const biMaxSafe = BigInt(Number.MAX_SAFE_INTEGER);

    test.each`
        data                                                                                            | options
        ${undefined}                                                                                    | ${undefined}
        ${'string'}                                                                                     | ${undefined}
        ${1}                                                                                            | ${undefined}
        ${1.1}                                                                                          | ${undefined}
        ${null}                                                                                         | ${undefined}
        ${true}                                                                                         | ${undefined}
        ${false}                                                                                        | ${undefined}
        ${[]}                                                                                           | ${undefined}
        ${[1, 2]}                                                                                       | ${undefined}
        ${['apple', 'banana', 'apple', 'banana', 'apple', 'pineapple']}                                 | ${undefined}
        ${['apple pie', 'apple', 'banana', 'apple-banana']}                                             | ${undefined}
        ${new Set(['apple', 'banana', 'pineapple'])}                                                    | ${undefined}
        ${new Set(['pineapple', 'apple', 'banana'])}                                                    | ${undefined}
        ${[new Set(['a', 'b', 'c']), new Set(['a', 'b', 'c']), new Set(['a', 'b', 'c'])]}               | ${undefined}
        ${new Map([['apple', 1], ['banana', 2], ['pineapple', 3]])}                                     | ${undefined}
        ${[new Map([['a', 1], ['b', 2], ['p', 3]]), new Map([['a', 1], ['b', 2], ['p', 3]])]}           | ${undefined}
        ${{}}                                                                                           | ${undefined}
        ${[{}, {}, {}]}                                                                                 | ${undefined}
        ${{ a: 1 }}                                                                                     | ${undefined}
        ${{ a: { b: 1 } }}                                                                              | ${undefined}
        ${{ a: { a: 'a', b: 42 } }}                                                                     | ${undefined}
        ${{ a: [1] }}                                                                                   | ${undefined}
        ${{ values: ['apple', 'banana', 'pineapple'], set: new Set(['apple', 'banana', 'pineapple']) }} | ${undefined}
        ${[{ a: 'a', b: 'b' }, { a: 'c', b: 'd' }, { b: 'b', a: 'a' }, ['a', 'b'], ['c', 'd']]}         | ${undefined}
        ${[{ a: 'a', b: 'b' }, { a: 'a', b: 'b' }, { a: 'a', b: 'b' }, { a: 'a', b: 'b' }]}             | ${{ dedupe: false }}
        ${sampleNestedData()}                                                                           | ${undefined}
        ${sampleRepeatedStrings()}                                                                      | ${undefined}
        ${/[\p{L}\p{M}]+/gu}                                                                            | ${undefined}
        ${[/[\p{L}\p{M}]+/gu, /[\p{L}\p{M}]+/gu, /[\p{Lu}\p{M}]+/gu]}                                   | ${undefined}
        ${[new Date('2024-01-01'), new Date('2024-01-01'), new Date('2024-01-02')]}                     | ${undefined}
        ${[1n, 2n, 1n, 2n, biMaxSafe, -biMaxSafe, biMaxSafe + 1n, -biMaxSafe - 1n]}                     | ${undefined}
        ${[Object(1n), Object('hello'), Object(/\w+/g), Object(null), Object([]), Object('hello')]}     | ${undefined}
    `('dehydrate $data $options', ({ data, options }) => {
        const v = toJSON(data, { dedupe: options?.dedupe });
        expect(v).toMatchSnapshot();
        expect(fromJSON(v)).toEqual(data);
        expect(fromJSON(JSON.parse(JSON.stringify(v)))).toEqual(data);
        expect(fromJSON(JSON.parse(stringify(data)))).toEqual(data);
        expect(fromJSON(JSON.parse(stringify(data, false)))).toEqual(data);
    });

    test.each`
        name             | data                             | options
        ${'fileList'}    | ${sampleFiles}                   | ${undefined}
        ${'fileObjects'} | ${await sampleFileListObjects()} | ${undefined}
    `('dehydrate $data $options', async ({ name, data, options }) => {
        const v = toJSON(data, { dedupe: options?.dedupe });
        await expect(stringifyFlatpacked(v)).toMatchFileSnapshot(`__snapshots__/${baseFilename}_${name}.jsonc`);
        await expect(JSON.stringify(v) + '\n').toMatchFileSnapshot(`__snapshots__/${baseFilename}_${name}.json`);
        await expect(JSON.stringify(data) + '\n').toMatchFileSnapshot(
            `__snapshots__/${baseFilename}_${name}.data.json`,
        );
        expect(fromJSON(v)).toEqual(data);
    });

    test("make sure dedupe doesn't break Sets", () => {
        const data = sampleNestedData();
        const value = { ...data, s: new Set(data.n) };

        const v = toJSON(value, { dedupe: true });
        const hv = fromJSON(v) as typeof value;
        expect(hv.s).toEqual(value.s);
        expect(hv.n).toEqual(value.n);
    });

    test("make sure dedupe doesn't break Maps", () => {
        const data = sampleNestedData();
        const value = { ...data, m: new Map(data.n.map((a) => [a, a])) };

        const v = toJSON(value, { dedupe: true });
        const hv = fromJSON(v) as typeof value;
        expect(hv.m).toEqual(value.m);
        expect(hv.n).toEqual(value.n);
    });

    const shortSampleFiles = sampleFiles.slice(0, 20);

    test.each`
        data                         | updated
        ${undefined}                 | ${undefined}
        ${'string'}                  | ${'string'}
        ${'string'}                  | ${'string + more'}
        ${['a', 'b', 'a', 'b']}      | ${['a', 'b', 'a', 'b', 'c']}
        ${['a', 'b', 'a', 'b', 'c']} | ${['a']}
        ${{}}                        | ${{ a: 'a' }}
        ${{ a: 1 }}                  | ${{ a: 1, b: 1 }}
        ${{ a: { b: 1 } }}           | ${{ a: { b: 1 }, b: { a: 1 } }}
        ${{ a: { a: 'a', b: 42 } }}  | ${{ a: { a: 'a', b: 42 }, b: 42 }}
        ${{ a: [1] }}                | ${{ a: [1, 2, 3] }}
        ${shortSampleFiles}          | ${[...shortSampleFiles.slice(0, 10), ...shortSampleFiles.slice(15)]}
        ${shortSampleFiles}          | ${[...shortSampleFiles.slice(0, 10), ...shortSampleFiles.slice(15), ...shortSampleFiles.slice(10, 15)]}
    `('Flatpack diff $data, $updated', ({ data, updated }) => {
        const fp = new FlatpackStore(data);
        const s0 = fp.stringify();
        fp.setValue(updated);
        expect(fromJSON(fp.toJSON())).toEqual(updated);
        const s1 = fp.stringify();
        const diff = createPatch('data', s0, s1);
        expect(diff).toMatchSnapshot();
    });
});

async function sampleFileList() {
    const data = await readFile(urlFileList, 'utf8');
    const files = data.split('\n');
    return files;
}

async function sampleFileListObjects() {
    const list = await sampleFileList();
    return list.map((filename) => ({ filename, fileType: getFileType(filename) }));
}

function getFileType(filename: string) {
    const baseName = filename.split('/').slice(-1).join('');
    return findMatchingFileTypes(baseName);
}

function sampleRepeatedStrings() {
    const fruit = ['apple', 'banana', 'apple', 'banana', 'apple', 'pineapple'];
    const sentence = 'There is a bit of fruit on the table. Some banana, apple, and pineapple.';
    const joinFruit = fruit.join('-');
    return {
        fruit,
        joinFruit,
        sentence,
    };
}

function sampleNestedData() {
    const a = { a: 'a', b: 'b' };
    const b = { a: 'c', b: 'd' };
    const r = /[\p{L}\p{M}]+/gu;
    const n = [a, b, { b: 'b', a: 'a' }, ['a', 'b'], ['c', 'd'], r, { r, rr: new RegExp(r) }];
    const s = new Set(n);
    const m = new Map([
        ['a', 'a'],
        ['b', 'b'],
    ]);

    const values = ['apple', 'banana', 'pineapple'];
    const rValues = [...values].reverse();
    const cValues = [...values];

    return {
        a,
        b,
        n,
        nn: n,
        nnn: [...n],
        s,
        r,
        m,
        ss: s,
        mm: m,
        values,
        rValues,
        cValues,
    };
}
