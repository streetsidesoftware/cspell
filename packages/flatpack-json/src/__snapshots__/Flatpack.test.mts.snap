// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Flatpack > Flatpack diff 'string', 'string + more' 1`] = `
"Index: data
===================================================================
--- data
+++ data
@@ -1,4 +1,4 @@
 [
 "Dehydrated JSON v1",
-"string"
+"string + more"
 ]
"
`;

exports[`Flatpack > Flatpack diff 'string', 'string' 1`] = `
"Index: data
===================================================================
--- data
+++ data
"
`;

exports[`Flatpack > Flatpack diff [ 'a', 'b', 'a', 'b' ], [ 'a', 'b', 'a', 'b', 'c' ] 1`] = `
"Index: data
===================================================================
--- data
+++ data
@@ -1,6 +1,7 @@
 [
 "Dehydrated JSON v1",
-[0,2,3,2,3],
+[0,2,3,2,3,4],
 "a",
-"b"
+"b",
+"c"
 ]
"
`;

exports[`Flatpack > Flatpack diff [ 'a', 'b', 'a', 'b', 'c' ], [ 'a' ] 1`] = `
"Index: data
===================================================================
--- data
+++ data
@@ -1,7 +1,5 @@
 [
 "Dehydrated JSON v1",
-[0,2,3,2,3,4],
-"a",
-"b",
-"c"
+[0,2],
+"a"
 ]
"
`;

exports[`Flatpack > Flatpack diff [ 'api/api.d.ts', 'dist/esm/Cache/cspell.cache.d.ts', …(18) ], [ 'api/api.d.ts', 'dist/esm/Cache/cspell.cache.d.ts', …(13) ] 1`] = `
"Index: data
===================================================================
--- data
+++ data
@@ -1,7 +1,7 @@
 [
 "Dehydrated JSON v1",
-[0,2,3,4,7,10,12,15,17,20,22,24,26,28,31,33,35,37,39,42,44],
+[0,2,3,4,7,10,12,15,17,20,22,35,37,39,42,44],
 "api/api.d.ts",
 "dist/esm/Cache/cspell.cache.d.ts",
 [2,5,6],
 [3,3,28],
@@ -22,17 +22,9 @@
 [2,21,6],
 [3,17,30],
 [2,16,23],
 "normalizeLanguageIds.d.ts",
-[2,16,25],
-"resolveDocument.d.ts",
-[2,13,27],
-"FeatureFlags/FeatureFalgs.test.d.ts",
-[2,29,30],
-[3,26,30],
-"lags.d.ts",
-[2,32,9],
-[3,26,22],
+0,0,0,0,0,0,0,0,0,
 [2,13,34],
 [3,23,16,9],
 [2,36,6],
 [3,33,21],
"
`;

exports[`Flatpack > Flatpack diff [ 'api/api.d.ts', 'dist/esm/Cache/cspell.cache.d.ts', …(18) ], [ 'api/api.d.ts', 'dist/esm/Cache/cspell.cache.d.ts', …(18) ] 1`] = `
"Index: data
===================================================================
--- data
+++ data
@@ -1,7 +1,7 @@
 [
 "Dehydrated JSON v1",
-[0,2,3,4,7,10,12,15,17,20,22,24,26,28,31,33,35,37,39,42,44],
+[0,2,3,4,7,10,12,15,17,20,22,35,37,39,42,44,24,26,28,31,33],
 "api/api.d.ts",
 "dist/esm/Cache/cspell.cache.d.ts",
 [2,5,6],
 [3,3,28],
"
`;

exports[`Flatpack > Flatpack diff { a: [ 1 ] }, { a: [ 1, 2, 3 ] } 1`] = `
"Index: data
===================================================================
--- data
+++ data
@@ -3,7 +3,9 @@
 [1,2,4],
 [0,3],
 "a",
 [0,5],
-[0,6],
-1
+[0,6,7,8],
+1,
+2,
+3
 ]
"
`;

exports[`Flatpack > Flatpack diff { a: { a: 'a', b: 42 } }, { a: { a: 'a', b: 42 }, b: 42 } 1`] = `
"Index: data
===================================================================
--- data
+++ data
@@ -1,10 +1,10 @@
 [
 "Dehydrated JSON v1",
-[1,2,4],
-[0,3],
+[1,6,2],
+[0,5,9],
 "a",
-[0,5],
+0,
 [1,6,8],
 [0,3,7],
 "b",
 [0,3,9],
"
`;

exports[`Flatpack > Flatpack diff { a: { b: 1 } }, { a: { b: 1 }, b: { a: 1 } } 1`] = `
"Index: data
===================================================================
--- data
+++ data
@@ -1,12 +1,14 @@
 [
 "Dehydrated JSON v1",
-[1,2,4],
+[1,4,10],
 [0,3],
 "a",
-[0,5],
+[0,3,7],
 [1,6,8],
 [0,7],
 "b",
 [0,9],
-1
+1,
+[0,5,11],
+[1,2,8]
 ]
"
`;

exports[`Flatpack > Flatpack diff { a: 1 }, { a: 1, b: 1 } 1`] = `
"Index: data
===================================================================
--- data
+++ data
@@ -1,8 +1,9 @@
 [
 "Dehydrated JSON v1",
-[1,2,4],
-[0,3],
+[1,2,6],
+[0,3,4],
 "a",
-[0,5],
-1
+"b",
+1,
+[0,5,5]
 ]
"
`;

exports[`Flatpack > Flatpack diff {}, { a: 'a' } 1`] = `
"Index: data
===================================================================
--- data
+++ data
@@ -1,4 +1,6 @@
 [
 "Dehydrated JSON v1",
-[1,0,0]
+[1,2,2],
+[0,3],
+"a"
 ]
"
`;

exports[`Flatpack > Flatpack diff undefined, undefined 1`] = `
"Index: data
===================================================================
--- data
+++ data
"
`;

exports[`Flatpack > dehydrate 'string' undefined 1`] = `
[
  "Dehydrated JSON v1",
  "string",
]
`;

exports[`Flatpack > dehydrate /[\\p{L}\\p{M}]+/gu undefined 1`] = `
[
  "Dehydrated JSON v1",
  [
    6,
    2,
    3,
  ],
  "[\\p{L}\\p{M}]+",
  "gu",
]
`;

exports[`Flatpack > dehydrate [ 'apple pie', 'apple', 'banana', 'apple-banana' ] undefined 1`] = `
[
  "Dehydrated JSON v1",
  [
    0,
    2,
    3,
    4,
    5,
  ],
  "apple pie",
  "apple",
  "banana",
  "apple-banana",
]
`;

exports[`Flatpack > dehydrate [ 'apple', 'banana', 'apple', 'banana', 'apple', 'pineapple' ] undefined 1`] = `
[
  "Dehydrated JSON v1",
  [
    0,
    2,
    3,
    2,
    3,
    2,
    4,
  ],
  "apple",
  "banana",
  "pineapple",
]
`;

exports[`Flatpack > dehydrate [ /[\\p{L}\\p{M}]+/gu, /[\\p{L}\\p{M}]+/gu, /[\\p{Lu}\\p{M}]+/gu ] undefined 1`] = `
[
  "Dehydrated JSON v1",
  [
    0,
    2,
    5,
    6,
  ],
  [
    6,
    3,
    4,
  ],
  "[\\p{L}\\p{M}]+",
  "gu",
  [
    6,
    3,
    4,
  ],
  [
    6,
    7,
    4,
  ],
  "[\\p{Lu}\\p{M}]+",
]
`;

exports[`Flatpack > dehydrate [ { a: 'a', b: 'b' }, { a: 'a', b: 'b' }, { a: 'a', b: 'b' }, …(1) ] { dedupe: false } 1`] = `
[
  "Dehydrated JSON v1",
  [
    0,
    2,
    6,
    7,
    8,
  ],
  [
    1,
    3,
    3,
  ],
  [
    0,
    4,
    5,
  ],
  "a",
  "b",
  [
    1,
    3,
    3,
  ],
  [
    1,
    3,
    3,
  ],
  [
    1,
    3,
    3,
  ],
]
`;

exports[`Flatpack > dehydrate [ { a: 'a', b: 'b' }, { a: 'c', b: 'd' }, { b: 'b', a: 'a' }, …(2) ] undefined 1`] = `
[
  "Dehydrated JSON v1",
  [
    0,
    2,
    6,
    2,
    3,
    7,
  ],
  [
    1,
    3,
    3,
  ],
  [
    0,
    4,
    5,
  ],
  "a",
  "b",
  [
    1,
    3,
    7,
  ],
  [
    0,
    8,
    9,
  ],
  "c",
  "d",
]
`;

exports[`Flatpack > dehydrate [ {}, {}, {} ] undefined 1`] = `
[
  "Dehydrated JSON v1",
  [
    0,
    2,
    2,
    2,
  ],
  [
    1,
    0,
    0,
  ],
]
`;

exports[`Flatpack > dehydrate [ 1, 2 ] undefined 1`] = `
[
  "Dehydrated JSON v1",
  [
    0,
    2,
    3,
  ],
  1,
  2,
]
`;

exports[`Flatpack > dehydrate [ 1n, 'hello', /\\w+/g, {}, [], 'hello' ] undefined 1`] = `
[
  "Dehydrated JSON v1",
  [
    0,
    2,
    5,
    7,
    10,
    11,
    12,
  ],
  [
    1,
    0,
    3,
  ],
  [
    8,
    4,
  ],
  1,
  [
    1,
    0,
    6,
  ],
  "hello",
  [
    6,
    8,
    9,
  ],
  "\\w+",
  "g",
  [
    1,
    0,
    0,
  ],
  [
    0,
  ],
  [
    1,
    0,
    6,
  ],
]
`;

exports[`Flatpack > dehydrate [ 1n, 2n, 1n, 2n, 9007199254740991n, -9007199254740991n, …(2) ] undefined 1`] = `
[
  "Dehydrated JSON v1",
  [
    0,
    2,
    4,
    2,
    4,
    6,
    8,
    10,
    12,
  ],
  [
    8,
    3,
  ],
  1,
  [
    8,
    5,
  ],
  2,
  [
    8,
    7,
  ],
  9007199254740991,
  [
    8,
    9,
  ],
  -9007199254740991,
  [
    8,
    11,
  ],
  "9007199254740992",
  [
    8,
    13,
  ],
  [
    2,
    14,
    11,
  ],
  "-",
]
`;

exports[`Flatpack > dehydrate [ Array(3) ] undefined 1`] = `
[
  "Dehydrated JSON v1",
  [
    0,
    2,
    3,
    4,
  ],
  [
    7,
    1704067200000,
  ],
  [
    7,
    1704067200000,
  ],
  [
    7,
    1704153600000,
  ],
]
`;

exports[`Flatpack > dehydrate [ Map{ 'a' => 1, 'b' => 2, 'p' => 3 }, Map{ 'a' => 1, 'b' => 2, 'p' => 3 } ] undefined 1`] = `
[
  "Dehydrated JSON v1",
  [
    0,
    2,
    2,
  ],
  [
    5,
    3,
    7,
  ],
  [
    0,
    4,
    5,
    6,
  ],
  "a",
  "b",
  "p",
  [
    0,
    8,
    9,
    10,
  ],
  1,
  2,
  3,
]
`;

exports[`Flatpack > dehydrate [ Set{ 'a', 'b', 'c' }, Set{ 'a', 'b', 'c' }, Set{ 'a', 'b', 'c' } ] undefined 1`] = `
[
  "Dehydrated JSON v1",
  [
    0,
    2,
    2,
    2,
  ],
  [
    4,
    3,
  ],
  [
    0,
    4,
    5,
    6,
  ],
  "a",
  "b",
  "c",
]
`;

exports[`Flatpack > dehydrate [] undefined 1`] = `
[
  "Dehydrated JSON v1",
  [
    0,
  ],
]
`;

exports[`Flatpack > dehydrate { a: [ 1 ] } undefined 1`] = `
[
  "Dehydrated JSON v1",
  [
    1,
    2,
    4,
  ],
  [
    0,
    3,
  ],
  "a",
  [
    0,
    5,
  ],
  [
    0,
    6,
  ],
  1,
]
`;

exports[`Flatpack > dehydrate { a: { a: 'a', b: 'b' }, b: { a: 'c', b: 'd' }, …(11) } undefined 1`] = `
[
  "Dehydrated JSON v1",
  [
    1,
    2,
    16,
  ],
  [
    0,
    3,
    4,
    5,
    6,
    7,
    8,
    9,
    10,
    11,
    12,
    13,
    14,
    15,
  ],
  "a",
  "b",
  "cValues",
  "m",
  "mm",
  "n",
  "nn",
  "nnn",
  "r",
  "rValues",
  "s",
  "ss",
  "values",
  [
    0,
    17,
    19,
    23,
    27,
    27,
    28,
    28,
    28,
    29,
    37,
    38,
    38,
    23,
  ],
  [
    1,
    18,
    18,
  ],
  [
    0,
    3,
    4,
  ],
  [
    1,
    18,
    20,
  ],
  [
    0,
    21,
    22,
  ],
  "c",
  "d",
  [
    0,
    24,
    25,
    26,
  ],
  "apple",
  "banana",
  "pineapple",
  [
    5,
    18,
    18,
  ],
  [
    0,
    17,
    19,
    17,
    18,
    20,
    29,
    32,
  ],
  [
    6,
    30,
    31,
  ],
  "[\\p{L}\\p{M}]+",
  "gu",
  [
    1,
    33,
    35,
  ],
  [
    0,
    11,
    34,
  ],
  "rr",
  [
    0,
    29,
    36,
  ],
  [
    6,
    30,
    31,
  ],
  [
    0,
    26,
    25,
    24,
  ],
  [
    4,
    39,
  ],
  [
    0,
    17,
    19,
    40,
    18,
    20,
    29,
    32,
  ],
  [
    1,
    18,
    18,
  ],
]
`;

exports[`Flatpack > dehydrate { a: { a: 'a', b: 42 } } undefined 1`] = `
[
  "Dehydrated JSON v1",
  [
    1,
    2,
    4,
  ],
  [
    0,
    3,
  ],
  "a",
  [
    0,
    5,
  ],
  [
    1,
    6,
    8,
  ],
  [
    0,
    3,
    7,
  ],
  "b",
  [
    0,
    3,
    9,
  ],
  42,
]
`;

exports[`Flatpack > dehydrate { a: { b: 1 } } undefined 1`] = `
[
  "Dehydrated JSON v1",
  [
    1,
    2,
    4,
  ],
  [
    0,
    3,
  ],
  "a",
  [
    0,
    5,
  ],
  [
    1,
    6,
    8,
  ],
  [
    0,
    7,
  ],
  "b",
  [
    0,
    9,
  ],
  1,
]
`;

exports[`Flatpack > dehydrate { a: 1 } undefined 1`] = `
[
  "Dehydrated JSON v1",
  [
    1,
    2,
    4,
  ],
  [
    0,
    3,
  ],
  "a",
  [
    0,
    5,
  ],
  1,
]
`;

exports[`Flatpack > dehydrate { fruit: [ 'apple', 'banana', 'apple', 'banana', 'apple', 'pineapple' ], …(2) } undefined 1`] = `
[
  "Dehydrated JSON v1",
  [
    1,
    2,
    6,
  ],
  [
    0,
    3,
    4,
    5,
  ],
  "fruit",
  "joinFruit",
  "sentence",
  [
    0,
    7,
    11,
    12,
  ],
  [
    0,
    8,
    9,
    8,
    9,
    8,
    10,
  ],
  "apple",
  "banana",
  "pineapple",
  "apple-banana-apple-banana-apple-pineapple",
  "There is a bit of fruit on the table. Some banana, apple, and pineapple.",
]
`;

exports[`Flatpack > dehydrate { values: [ 'apple', 'banana', 'pineapple' ], …(1) } undefined 1`] = `
[
  "Dehydrated JSON v1",
  [
    1,
    2,
    5,
  ],
  [
    0,
    3,
    4,
  ],
  "set",
  "values",
  [
    0,
    6,
    7,
  ],
  [
    4,
    7,
  ],
  [
    0,
    8,
    9,
    10,
  ],
  "apple",
  "banana",
  "pineapple",
]
`;

exports[`Flatpack > dehydrate {} undefined 1`] = `
[
  "Dehydrated JSON v1",
  [
    1,
    0,
    0,
  ],
]
`;

exports[`Flatpack > dehydrate 1 undefined 1`] = `
[
  "Dehydrated JSON v1",
  1,
]
`;

exports[`Flatpack > dehydrate 1.1 undefined 1`] = `
[
  "Dehydrated JSON v1",
  1.1,
]
`;

exports[`Flatpack > dehydrate Map{ 'apple' => 1, 'banana' => 2, 'pineapple' => 3 } undefined 1`] = `
[
  "Dehydrated JSON v1",
  [
    5,
    2,
    6,
  ],
  [
    0,
    3,
    4,
    5,
  ],
  "apple",
  "banana",
  "pineapple",
  [
    0,
    7,
    8,
    9,
  ],
  1,
  2,
  3,
]
`;

exports[`Flatpack > dehydrate Set{ 'apple', 'banana', 'pineapple' } undefined 1`] = `
[
  "Dehydrated JSON v1",
  [
    4,
    2,
  ],
  [
    0,
    3,
    4,
    5,
  ],
  "apple",
  "banana",
  "pineapple",
]
`;

exports[`Flatpack > dehydrate Set{ 'pineapple', 'apple', 'banana' } undefined 1`] = `
[
  "Dehydrated JSON v1",
  [
    4,
    2,
  ],
  [
    0,
    3,
    4,
    5,
  ],
  "pineapple",
  "apple",
  "banana",
]
`;

exports[`Flatpack > dehydrate false undefined 1`] = `
[
  "Dehydrated JSON v1",
  false,
]
`;

exports[`Flatpack > dehydrate null undefined 1`] = `
[
  "Dehydrated JSON v1",
  null,
]
`;

exports[`Flatpack > dehydrate true undefined 1`] = `
[
  "Dehydrated JSON v1",
  true,
]
`;

exports[`Flatpack > dehydrate undefined undefined 1`] = `
[
  "Dehydrated JSON v1",
]
`;
